rewrite-test/src/main/java/org/openrewrite/test/RewriteTest.java

```diff
@@ -551,7 +551,7 @@ default void validateRecipeOptions(Recipe recipe) {
 class RewriteTestUtils {
     static boolean groupSourceSpecsByParser(List<Parser.Builder> parserBuilders, Map<Parser.Builder, List<SourceSpec<?>>> sourceSpecsByParser, SourceSpec<?> sourceSpec) {
         for (Map.Entry<Parser.Builder, List<SourceSpec<?>>> entry : sourceSpecsByParser.entrySet()) {
-            if (entry.getKey().getSourceFileType().equals(sourceSpec.sourceFileType)) {
+            if (entry.getKey().getSourceFileType().equals(sourceSpec.sourceFileType) && sourceSpec.getParser().getClass().isAssignableFrom(entry.getKey().getClass())) {
```

Comment:
The reason the bug can happen in the first place is that both `MavenParser.accept()` and `XmlParser.accept()` will return `true` for a file with an xml extension. And both produce `Xml.Document`... the only difference in the result is that the maven parser also attaches our maven marker with dependency resolution info as metadata.

The test framework will stop trying to match sources to parsers once it finds a parser that will accept the test spec. It isn't clear to me looking at this that it would address that issue.

Reply:
There is a `Parser$Builder` on the `SourceSpec`, the `Parser$Builder` for the `pomXml` is a `MavenParser$Builder` while the `Parser$Builder` for the `xml` is a `XmlParser$Builder`. I don't know how to test this within RewriteTestTest (because it relies on rewrite-maven which is not on the classpath), but in my recipe in rewrite-java-security, I verified this change will use the correct Parser for both the `xml` and the `pomXml`.

Reply:
I could put a test in rewrite-maven ... let me try that

Reply:
https://github.com/openrewrite/rewrite/pull/2852/files#diff-de2491affc07fd204feb64109a63b15e34034a5c40490293d1c712e11bd2b9a5R20

Verified that without my change to `RewriteTest`, I get an error. With my change to `RewriteTest`, the test succeeds.